src/services/db/__tests__/GitFileSystemService.spec.ts

```diff
@@ -677,182 +812,224 @@ describe("GitFileSystemService", () => {
-        .sort((a, b) => a.name.localeCompare(b.name))
-
-      expect(actual).toMatchObject([expectedAnotherFakeFile, expectedFakeDir])
+      expect(actual._unsafeUnwrapErr()).toBeInstanceOf(GitFileSystemError)
+      expect(spyRollback).toHaveBeenCalledWith("fake-repo", "test-commit-sha")
     })
+  })
 
-    it("should return an empty result if the directory contain files that are all untracked", async () => {
-      MockSimpleGit.cwd.mockReturnValueOnce({
-        revparse: jest.fn().mockRejectedValueOnce(new GitError()),
-      })
-      MockSimpleGit.cwd.mockReturnValueOnce({
-        revparse: jest.fn().mockRejectedValueOnce(new GitError()),
-      })
+  describe("read", () => {
+    it("should read the contents of a file successfully", async () => {
       MockSimpleGit.cwd.mockReturnValueOnce({
-        revparse: jest.fn().mockRejectedValueOnce(new GitError()),
+        revparse: jest.fn().mockResolvedValueOnce("fake-hash"),
       })
 
-      const actual = await GitFileSystemService.listDirectoryContents(
+      const expected: GitFile = {
+        content: "fake content",
+        sha: "fake-hash",
+      }
+      const actual = await GitFileSystemService.read(
         "fake-repo",
-        ""
+        "fake-dir/fake-file"
       )
 
-      expect(actual._unsafeUnwrap()).toHaveLength(0)
+      expect(actual._unsafeUnwrap()).toEqual(expected)
     })
 
-    it("should return an empty result if the directory is empty", async () => {
-      const actual = await GitFileSystemService.listDirectoryContents(
-        "fake-repo",
-        "fake-empty-dir"
-      )
-
-      expect(actual._unsafeUnwrap()).toHaveLength(0)
-    })
+    it("should return a NotFoundError if the file does not exist", async () => {
+      MockSimpleGit.cwd.mockReturnValueOnce({
+        revparse: jest.fn().mockRejectedValueOnce(new GitError()),
+      })
 
-    it("should return a GitFileSystemError if the path is not a directory", async () => {
-      const result = await GitFileSystemService.listDirectoryContents(
+      const result = await GitFileSystemService.read(
         "fake-repo",
-        "fake-dir/fake-file"
+        "fake-dir/non-existent-file"
       )
 
-      expect(result._unsafeUnwrapErr()).toBeInstanceOf(GitFileSystemError)
+      expect(result._unsafeUnwrapErr()).toBeInstanceOf(NotFoundError)
     })
 
-    it("should return a NotFoundError if the path does not exist", async () => {
-      const result = await GitFileSystemService.listDirectoryContents(
+    it("should return a error if an error occurred when getting the Git blob hash", async () => {
+      MockSimpleGit.cwd.mockReturnValueOnce({
+        revparse: jest.fn().mockRejectedValueOnce(new GitError()),
+      })
+
+      const result = await GitFileSystemService.read(
         "fake-repo",
-        "non-existent-dir"
+        "fake-dir/fake-file"
       )
 
-      expect(result._unsafeUnwrapErr()).toBeInstanceOf(NotFoundError)
+      expect(result.isErr()).toBeTrue()
```

Comment:
how about tests to ensure rollbacks are done correctly - i.e. when failures occur along one of the many steps in create flow

Reply:
Hmm we actually only rollback in either the actual file creation, or when committing the change. I'm not entirely sure how to mock the fs call itself!

Reply:
okay, let's focus on the manual testing for rollback and add this in later. I think @dcshzj added a mock-fs library into the spec file which we can use possibly